File: java/src/leetcode/editor/cn/367-ValidPerfectSquare.java
package leetcode.editor.cn;

/*
 * @lc app=leetcode.cn id=367 lang=java
 *
 * [367] Valid Perfect Square
 *
 * https://leetcode-cn.com/problems/valid-perfect-square/description/
 *
 * algorithms
 * Easy (44.85%)
 * Likes:    420
 * Dislikes: 0
 * Total Accepted:    174K
 * Total Submissions: 388K
 * Testcase Example:  '16'
 *
 * Given a positive integer num, write a function which returns True if num is
 * a perfect square else False.
 * 
 * Follow up: Do not use any built-in library function such as sqrt.
 * 
 * 
 * Example 1:
 * Input: num = 16
 * Output: true
 * Example 2:
 * Input: num = 14
 * Output: false
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= num <= 2^31 - 1
 * 
 * 
 */

// @lc code=start
class Solution {
    public boolean isPerfectSquare(int num) {
        int low = 1, high = num;
        while (low <= high) {
            int mid = (low + high) >> 1;

            long mid2 = (long) mid * mid;
            if (mid2 == num) {
                return true;
            } else if (mid2 < num) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return false;
    }
}
// @lc code=end

File: javascript/src/leetcode/editor/cn/367-ValidPerfectSquare.js
/*
 * @lc app=leetcode.cn id=367 lang=javascript
 *
 * [367] 有效的完全平方数
 *
 * https://leetcode-cn.com/problems/valid-perfect-square/description/
 *
 * algorithms
 * Easy (43.55%)
 * Likes:    178
 * Dislikes: 0
 * Total Accepted:    48K
 * Total Submissions: 110.3K
 * Testcase Example:  '16'
 *
 * 给定一个正整数 num，编写一个函数，如果 num 是一个完全平方数，则返回 True，否则返回 False。
 * 
 * 说明：不要使用任何内置的库函数，如  sqrt。
 * 
 * 示例 1：
 * 
 * 输入：16
 * 输出：True
 * 
 * 示例 2：
 * 
 * 输入：14
 * 输出：False
 * 
 * 
 */

// @lc code=start
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function (num) {
    if (num < 2) {
        return true
    }

    var left = 2,
        right = num / 2
    while (left <= right) {
        let x = parseInt(left + (right - left) / 2)
        let squared = x * x
        if (squared > num) {
            right = x - 1
        } else if (squared < num) {
            left = x + 1
        } else {
            return true
        }
    }

    return false
};
// @lc code=end